src/main/scala/com/github/mrpowers/spark/fast/tests/DataFrameComparer.scala

```diff
@@ -14,17 +14,24 @@ trait DataFrameComparer extends DatasetComparer {
       ignoreColumnNames: Boolean = false,
       orderedComparison: Boolean = true,
       ignoreColumnOrder: Boolean = false,
-      truncate: Int = 500
+      truncate: Int = 500,
   ): Unit = {
-    assertSmallDatasetEquality(
-      actualDF,
-      expectedDF,
-      ignoreNullable,
-      ignoreColumnNames,
-      orderedComparison,
-      ignoreColumnOrder,
-      truncate
-    )
+    SchemaComparer.assertSchemaEqual(actualDF, expectedDF, ignoreNullable, ignoreColumnNames, ignoreColumnOrder)
+    val actual = if (ignoreColumnOrder) orderColumns(actualDF, expectedDF) else actualDF
+    if (orderedComparison)
+      assertSmallDataFrameEquality(actual, expectedDF, truncate)
+    else
+      assertSmallDataFrameEquality(defaultSortDataset(actual), defaultSortDataset(expectedDF), truncate)
+  }
+
+  def assertSmallDataFrameEquality(actualDF: DataFrame, expectedDF: DataFrame, truncate: Int): Unit = {
+    val a = actualDF.collect()
+    val e = expectedDF.collect()
+    if (!a.toSeq.approximateSameElements(e, (o1: Row, o2: Row) => o1.equals(o2))) {
+      val arr = ("Actual Content", "Expected Content")
+      val msg = "Diffs\n" ++ DataframeUtil.showDataframeDiff(arr, a, e, truncate)
```

Comment:
maybe we should add an option on assertSmallDatasetEquality instead for people to switch to the clearer error message.
so both compare function benefit from it and also reduce duplication
What do you think @SemyonSinchenko, @happytomatoe ?

Reply:
@zeotuan Dataset knows nothing about the underlying type. Probably the only option is to use reflection for datasets and arrays. Using reflection is more complicated and probably messy. Will search for an example of an implementation.